### ISSUE TITLE
[V2] Theming doesn't work with reactstrap 4.2.0

### ISSUE THREAD

**Issue by L3ARNING2B** [styled-components/styled-components#574]
Hi everyone,

I have read this very recent [article](https://www.smashingmagazine.com/2017/01/styled-components-enforcing-best-practices-component-based-systems/) claiming styled-components is the future so I decided to give a try.

I want to to customize my [reactstrap](https://reactstrap.github.io/) app. _(React equivalent of Bootstrap 4)_ 

I don't wan't to import the bootstrap css stylesheet. I rather make my own styling based on the class used in bootstrap. This is where styled-components come from! I Hoped...

For some reason, the custom theme passed to the `<ThemeProvider theme={theme} />` disappear for the one I've configured in my component `Button.defaultProps` right after a re-rending due to state change.

I've looked into their [__Button__ components](https://github.com/reactstrap/reactstrap/blob/master/src/Button.js) and I still wonder, their components are fairly simple.

I have created a webpackbin to demonstrate the issue.

I have also opened an issue on reactstrap: https://github.com/reactstrap/reactstrap/issues/355

According to @TheSharpieOne

> Yeah, definitely looks like an issue on styled-components side. ... Seems like Styled(*) shouldn't be passing down the props it uses. [...]
IDK, but looking at this line in styled-components it looks like it only subscribes and doesn't set the initial state based on the [default] props.

## Version
styled-components 2.0.0-7
reactstrap 4.2.0

## Reproduction

https://www.webpackbin.com/bins/-KetejqQhRZrV5MyzPpr

## Steps to reproduce

Click on the button within the reproduction and see.

## Expected Behavior

Custom theme to be applied

## Actual Behavior

Default theme is applied

**Comment by L3ARNING2B** [comment]
I've opened an issue on reactstrap as well : https://github.com/reactstrap/reactstrap/issues/355

**Comment by kopax** [comment]
Thanks for reporting. Looks similar to #549

**Comment by santaclauze** [comment]
Yup, working on similar issues. Any leads would be grateful!

**Comment by santaclauze** [comment]
# Version

styled-components 2.0.0-7
reactstrap 4.2.0


# Steps to reproduce

https://www.webpackbin.com/bins/-Kf5Tdg7bfj2oWZzIb1V

Case A - Button: Click on one of the left buttons to reproduce the issue with the component Button.

Case B - ButtonClass: Click on one of the right buttons to reproduce the issue with the component ButtonClass.

Case C - ButtonBetter: Copy paste ButtonBetter component (as Webpackbin does not support the `static` keyword), uncomment line 11 `// static defaultProps = defaultProps;` to reproduce fix.

As you will see, the issue was resolved as follows:

Problematic Case:
```jsx
const defaultProps = {
  theme: themeBs,
};

class ButtonClass extends React.Component {
...
}

ButtonClass.defaultProps = defaultProps;
export default ButtonClass;
```

Fixed Case:
```jsx
const defaultProps = {
  theme: themeBs,
};

class ButtonBetter extends React.Component {
  static defaultProps = defaultProps;
...
}

export default ButtonBetter;
```
Cause for Fix: moving the defaultProps inside the component and removing the `Button.defaultProps = defaultProps;`

### Logs
Case A:  seem normal:
Case B:  seem normal:
Case C: returns undefined multiple times for `${(props) => console.log('ButtonBetter within styled', props.theme.main)}` but magically returns the correct value once. 

# Expected Behavior

Custom theme to be applied

# Actual Behavior

Default theme is applied

# Questions

1. Is there any way to use ` static` keyword in Webpackbin? Or is there any work around to produce an equivalent results?

2. Why does the logs for Case C return undefined, rather then correct value?

3. What will happen to stateless components and styled if we are obliged to define `defaultProps` within the component? (As it seems that you cannot define the defaultProps inside a stateless component)

**Comment by k15a** [comment]
Could you please check if this is also a bug in v1?

**Comment by k15a** [comment]
Ok this is a bug with v2 only. I think I have already fixed that bug in v1 so I will have to search for the PR.

**Comment by santaclauze** [comment]
PR does not seem to have been added in v.2, I tested the commit you added in the PR and I confirm that it resolves the issue.

Thanks for that!

**Comment by kopax** [comment]
Is it possible to refix this ? I can do the PR it's only 5 minutes. :+1:

**Comment by k15a** [comment]
Should be copy paste. 👍

**Comment by L3ARNING2B** [comment]
Many thanks for the fix added in v 2.0.0-9.

**Comment by nik72619c** [comment]
Hi guys,
I am using reactstrap (v6.4.0) and styled-components(v5.1.1) and using `ThemeProvider` to provide theme to all the styled-components in the app. But due to some reason, the theme prop is not available inside the styled-components created inside the reactstrap `Tooltip` component. Helps appreciated, thanks !
